test/integration/cds_integration_test.cc

```diff
@@ -0,0 +1,211 @@
+      // Setup cds and corresponding gRPC cluster.
+      auto* cds_config =
+          bootstrap.mutable_dynamic_resources()->mutable_cds_config()->mutable_api_config_source();
+      bootstrap.mutable_dynamic_resources()->clear_ads_config();
+      bootstrap.mutable_dynamic_resources()->clear_lds_config();
+
+      cds_config->set_api_type(envoy::api::v2::core::ApiConfigSource::GRPC);
+      cds_config->mutable_request_timeout()->set_seconds(1);
+      auto* grpc_service = cds_config->add_grpc_services();
+      setGrpcService(*grpc_service, "my_cds_cluster", fake_upstreams_[0]->localAddress());
+      grpc_service->mutable_envoy_grpc()->set_cluster_name("my_cds_cluster");
+
+      auto* cds_cluster = bootstrap.mutable_static_resources()->add_clusters();
+      cds_cluster->set_name("my_cds_cluster");
+      cds_cluster->mutable_connect_timeout()->set_seconds(5);
+      auto* sockaddr = cds_cluster->add_hosts()->mutable_socket_address();
+      sockaddr->set_protocol(envoy::api::v2::core::SocketAddress::TCP);
+      sockaddr->set_address(Network::Test::getLoopbackAddressString(ipVersion()));
+      sockaddr->set_port_value(0);
+      cds_cluster->clear_http2_protocol_options();
+    });
+    setUpstreamProtocol(FakeHttpConnection::Type::HTTP2);
+
+    HttpIntegrationTest::initialize();
+
+    fake_upstreams_[0]->set_allow_unexpected_disconnects(false);
+    // Causes cds_connection_ to be filled with a newly constructed FakeHttpConnection.
+    AssertionResult result =
+        fake_upstreams_[0]->waitForHttpConnection(*dispatcher_, cds_connection_);
+    RELEASE_ASSERT(result, result.message());
+    result = cds_connection_->waitForNewStream(*dispatcher_, cds_stream_);
+    RELEASE_ASSERT(result, result.message());
+    cds_stream_->startGrpcStream();
+
+    EXPECT_TRUE(compareDiscoveryRequest(Config::TypeUrl::get().Cluster, "", {}));
+    sendDiscoveryResponse<envoy::api::v2::Cluster>(Config::TypeUrl::get().Cluster,
+                                                   {buildCluster("cluster_0")}, "1");
+  }
+
+  testing::NiceMock<Server::Configuration::MockTransportSocketFactoryContext> factory_context_;
+  FakeStreamPtr cds_stream_;
+  FakeHttpConnectionPtr cds_connection_;
+  FakeHttpConnectionPtr upstream_connection_;
+};
+
+// GoogleGrpc causes problems.
+INSTANTIATE_TEST_CASE_P(IpVersionsClientType, CdsIntegrationTest,
+                        testing::Combine(testing::ValuesIn(TestEnvironment::getIpVersionsForTest()),
+                                         testing::Values(Grpc::ClientType::EnvoyGrpc)));
+
+// Tests Envoy HTTP health checking a single healthy endpoint and reporting that it is
+// indeed healthy to the server.
+TEST_P(CdsIntegrationTest, RouterRequestAndResponseWithBodyNoBuffer) {
+  // Controls how many fake_upstreams_.emplace_back(new FakeUpstream) will happen in
+  // BaseIntegrationTest::createUpstreams() (which is part of initialize()).
+  setUpstreamCount(1);
```

Comment:
How come this isn't 2, one for the CDS server and one for the dataplane upstream?

Reply:
I could only get things working when I had the CDS server itself be presented as the dataplane upstream as well. I would prefer being able to get a separate dataplane one. I figured this was good for a start, since the Envoy is still discovering an upstream HTTP server, even if a normal deployment would not have the CDS server itself as one of the dataplane upstreams.

Reply:
`ads_integration_test` has the dual upstreams, so I think this shouldn't be too hard to make work. It's a bit clearer than reusing the same upstream for control and data plane.

Reply:
Done!